llama_index/query_pipeline/query.py

```diff
@@ -0,0 +1,582 @@
+            **kwargs,
+        )
+
+        if chain is not None:
+            # generate implicit link between each item, add
+            self.add_chain(chain)
+
+    def add_chain(self, chain: Sequence[QUERY_COMPONENT_TYPE]) -> None:
+        """Add a chain of modules to the pipeline.
+
+        This is a special form of pipeline that is purely sequential/linear.
+        This allows a more concise way of specifying a pipeline.
+
+        """
+        # first add all modules
+        module_keys = []
+        for module in chain:
+            module_key = str(uuid.uuid4())
+            self.add(module_key, module)
+            module_keys.append(module_key)
+
+        # then add all links
+        for i in range(len(chain) - 1):
+            self.add_link(src=module_keys[i], dest=module_keys[i + 1])
+
+    def add_modules(self, module_dict: Dict[str, QUERY_COMPONENT_TYPE]) -> None:
+        """Add modules to the pipeline."""
+        for module_key, module in module_dict.items():
+            self.add(module_key, module)
+
+    def add(self, module_key: str, module: QUERY_COMPONENT_TYPE) -> None:
+        """Add a module to the pipeline."""
+        # if already exists, raise error
+        if module_key in self.module_dict:
+            raise ValueError(f"Module {module_key} already exists in pipeline.")
+
+        if isinstance(module, ChainableMixin):
+            module = module.as_query_component()
+        else:
+            pass
+
+        self.module_dict[module_key] = cast(QueryComponent, module)
+
+    def add_link(
+        self,
+        src: str,
+        dest: str,
+        src_key: Optional[str] = None,
+        dest_key: Optional[str] = None,
+    ) -> None:
+        """Add a link between two modules."""
+        if src not in self.module_dict:
+            raise ValueError(f"Module {src} does not exist in pipeline.")
+        if src not in self.edge_dict:
+            self.edge_dict[src] = []
+        self.edge_dict[src].append(
+            Link(src=src, dest=dest, src_key=src_key, dest_key=dest_key)
+        )
+
+    def _get_root_keys(self) -> List[str]:
+        """Get root keys."""
+        # first add all to root keys, then remove
+        root_keys: Set[str] = set(self.module_dict.keys())
+        # get all modules without upstream dependencies
+        for module_key in self.edge_dict:
+            for link in self.edge_dict[module_key]:
+                if link.dest in root_keys:
+                    root_keys.remove(link.dest)
+        return list(root_keys)
+
+    def _get_leaf_keys(self) -> List[str]:
+        """Get leaf keys."""
+        # get all modules without downstream dependencies
+        leaf_keys = []
+        for module_key in self.module_dict:
+            if module_key not in self.edge_dict or len(self.edge_dict[module_key]) == 0:
+                leaf_keys.append(module_key)
+        return list(leaf_keys)
+
+    def _ancestral_sort(self, queue: List[InputTup]) -> List[InputTup]:
```

Comment:
is this equivalent to topological sort? 

https://pelegm-networkx.readthedocs.io/en/latest/reference/generated/networkx.algorithms.dag.topological_sort.html

Reply:
seems like it